### PR TITLE
[JENKINS-58450] CHANGE_FORK is set when the fork name differs from the upstream name

### PR DESCRIPTION
**Pull request by v1v**
[JENKINS-58450](https://issues.jenkins-ci.org/browse/JENKINS-58450)

## Highlights
- `CHANGE_FORK` is now set to the owner/repo when the repo differs from the name of the upstream repo.

## Tests

- Used to be when even the repo name was different.

image

- Now

image

## Tasks
- [x] UTs
- [x] Confirm whether the changed constructor is required for backward compatibilities even though the implementation might differ.
- [x] Test it against the major trait/strategy plugins. List of usage for the class PullRequestSCMHead  [here](https://github.com/search?q=org%3Ajenkinsci+PullRequestSCMHead&type=Code)
  - https://github.com/jenkinsci/github-branch-source-plugin/blob/master/src/main/java/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait.java . I tested it within this PR.
  - https://github.com/jenkinsci/github-branch-source-plugin/blob/master/src/main/java/org/jenkinsci/plugins/github_branch_source/OriginPullRequestDiscoveryTrait.java . I tested it within this PR.
  - https://github.com/jenkinsci/scm-trait-commit-skip-plugin/blob/446f3153256ad9b8403aaba6d597559e8a6b9aaa/github-scm-trait-commit-skip/src/main/java/org/jenkinsci/plugins/scm_filter/GitHubCommitSkipTrait.java . See https://github.com/jenkinsci/github-branch-source-plugin/pull/235#issuecomment-513947855
- [x] Describe/show what the results of this change will be for consumers of CHANGE_FORK:
  -  https://github.com/jenkinsci/pipeline-model-definition-plugin . See https://github.com/jenkinsci/github-branch-source-plugin/pull/235#issuecomment-512972423
  - https://github.com/jenkinsci/branch-api-plugin  . See https://github.com/jenkinsci/github-branch-source-plugin/pull/235#issuecomment-512977353

### CODE REVIEW

**Comment by v1v** [comment]
@bitwiseman , not sure if you are the owner of this repo, as I've just found you managed to merge the last few PRs, if you don't mind, may I ask you for some reviews?

Thanks

**Comment by bitwiseman** [comment]
@v1v 
I'm a contributor, so I can help with reviews and such.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/PullRequestSCMHead.java [hunk 1]
```diff
@@ -87,9 +87,15 @@
...
         this.sourceBranch = pr.getHead().getRef();
-        this.origin = pr.getRepository().getOwnerName().equalsIgnoreCase(sourceOwner)
-                ? SCMHeadOrigin.DEFAULT
-                : new SCMHeadOrigin.Fork(this.sourceOwner);
+
+        if (pr.getRepository().getOwnerName().equalsIgnoreCase(sourceOwner)) {
+            this.origin = SCMHeadOrigin.DEFAULT;
+        } else {
+            // if the forked repo name differs from the upstream repo name
+            this.origin = pr.getHead().getRepository().getName().equalsIgnoreCase(repoName)
```
Comment (bitwiseman) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/PullRequestSCMHead.java]: I think you can do this without changing the constructor and so on:
```suggestion
            this.origin = pr.getHead().getRepository().getName().equalsIgnoreCase(pr.getBase().getRepository().getName())
```

If this works you can rollback the changes out calling code and constructor.

**Comment by bitwiseman** [comment]
@v1v 
First, great job including tests and commenting.  

I can see how the intended change makes sense.  In the case where the repo names differ, the `FORK` value will include the repo name.  

__Pros__
* Solves definite flaw in the way the plugin exposes fork information (it is not really sufficient in all reasonable cases)
* Use case is uncommon but not exceptional - It is uncommon for fork repos to have different names from their roots, but it is valid and not that rare. 
* Does not break existing cases (when repo names are the same)
* Changes the data stored in underlying object not just the exposed `CHANGE_FORK` - all consumers of `origin` will see the same information.
* What you've done here is in accordance with the API as described here: 
https://github.com/jenkinsci/scm-api-plugin/blob/5b5e58e70489167414b77fd1a7e8a92bf7b3e68e/src/main/java/jenkins/scm/api/SCMHeadOrigin.java#L129-L152

__Concerns__
* New value format is incompatible with old format - old format is `owner-name`, new format is `owner-name/repo-name`.  If I'm a consumer of `CHANGE_FORK`, I must now check if it contains a slash to know which format to use.  
* Changes the data stored in underlying object not just the exposed `CHANGE_FORK` - despite this being valid based on the API description, all consumer of `origin` will be effected and will need to be tested as correctly handling this change. 

For example, I'm pretty sure there are traits/strategies that depend on the value of `origin` being `owner-name` and may break if it were `owner-name/repo-name`.  That will need testing before we merge this. 

__Action items__:
* Integrate suggested code change (or reject it with reasons if I'm wrong about it)
* Use the incremental of this PR and test it against the major trait/strategy plugins. If there breaks, that will mean updating those plugins as well.
* Describe/show what the results of this change will be for consumers of `CHANGE_FORK`.  For most scenarios, this is a non-breaking change - and for those it does break most were already broken - but we still need to be clear about the results so we can describe them in release notes. 

Thanks again for this PR.  It is high quality and most of the way to mergable.  We just need a few more details and tweaks covered.  

Related notes: 
`CHANGE_FORK` is exposed here:
https://github.com/jenkinsci/branch-api-plugin/blob/c4d394415cf25b6890855a08360119313f1330d2/src/main/java/jenkins/branch/BranchNameContributor.java#L72-L74

**Comment by v1v** [comment]
That's a great plan. I'll proceed to validate the different plugins that consume this one:
- https://github.com/search?l=Maven+POM&q=org%3Ajenkinsci+github-branch-source&type=Code

I'll try firstly locally and likely create a PR to bump the version, using initially the incremental ones to validate it in the CI. Then, they can decide whether to upgrade once the release is out.

Thanks for the feedback

**Comment by v1v** [comment]
For the repo https://github.com/jenkinsci/pipeline-model-definition-plugin I tested it out the below pipeline:

```
pipeline {
    agent any
    stages {
        stage('Same repo name') {
            when { changeRequest fork: 'v1v' }
            steps {
                echo env.CHANGE_FORK
            }
        }
        stage('Different repo name') {
            when { changeRequest fork: 'v1v/pipeline-1' }
            steps {
                echo env.CHANGE_FORK
            }
        }
    }
}
```

## What did I do?
- git clone https://github.com/jenkinsci/pipeline-model-definition-plugin
- `mvn hpi:run`
- Install the plugin: https://repo.jenkins-ci.org/incrementals/org/jenkins-ci/plugins/github-branch-source/2.5.4-rc860.9c636841daa2/
- Install the declarative and pipeline plugins. Further details https://gist.github.com/v1v/306285c2e702b84c669788eef841b18c
- Create MBP project for https://github.com/super-mario-bros-pipelines/pipeline-2
- Create MBP project for https://github.com/super-mario-bros-pipelines/pipeline

**Comment by v1v** [comment]
For the repo https://github.com/jenkinsci/branch-api-plugin I tested it out the below pipeline:

```
pipeline {
    agent any
    stages {
        stage('Same repo name') {
            when { changeRequest fork: 'v1v' }
            steps {
                echo env.CHANGE_FORK
            }
        }
        stage('Different repo name') {
            when { changeRequest fork: 'v1v/pipeline-1' }
            steps {
                echo env.CHANGE_FORK
            }
        }
    }
}
```

## What did I do?
- git clone https://github.com/jenkinsci/branch-api-plugin
- `mvn hpi:run`
- Install the plugin: https://repo.jenkins-ci.org/incrementals/org/jenkins-ci/plugins/github-branch-source/2.5.4-rc860.9c636841daa2/
- Install the declarative and pipeline plugins. Further details https://gist.github.com/v1v/7330c2f32e10657682f69fc3d6a66e61
- Create MBP project for https://github.com/super-mario-bros-pipelines/pipeline-2
- Create MBP project for https://github.com/super-mario-bros-pipelines/pipeline

## Test output for pipeline repo
image

## Test output for the pipeline-2 repo

image

**Comment by v1v** [comment]
For the repo I tested:

- https://github.com/super-mario-bros-pipelines/pipeline/pull/3/commits/c725e70eb7e6e3bb7c9e188e6ff48b28154e326c 
 and I've got the expected behaviour, aka no builds for the PR.

image

- https://github.com/super-mario-bros-pipelines/pipeline/pull/1/commits/349eacefbc3727d41978e3fc5bda7fa35dd07ac0
 and I've got the expected behaviour, aka no builds for the PR.

image

**Comment by v1v** [comment]
@bitwiseman, I think I managed to test all the cases I could find where either the `CHANGE_FORK` or the `PullRequestSCMHead` was used/referenced. I also updated the description with the link to those tests, even though they have been added as comments, but might help to understand what I've done so far. Please let me know what you think.

Thanks

**Comment by bitwiseman** [comment]
There's been some comments added to https://issues.jenkins-ci.org/browse/JENKINS-58450.

**Pull request merged**